### PR TITLE
[192] Generalise public discourse analysis 

### LINKED ISSUE #192: Make public discourse analysis generalisable
Test the functions in [pd_analysis_utils](https://github.com/nestauk/innovation_sweet_spots/blob/dev/innovation_sweet_spots/utils/pd/pd_analysis_utils.py)  are usable for any input data that has:
- timestamp
- unique identifier
- text content

Make functions generalisable if they are not already.

Examples of usage can be found [here](https://github.com/nestauk/innovation_sweet_spots/tree/dev/innovation_sweet_spots/analysis/examples/public_discourse_analysis).

Test with Guardian and Hansard data.
Guardian data will need API setup.
Hansard test data can be found [here](https://docs.google.com/spreadsheets/d/1S5KGgZJtJvI-MU-kR649X4kIAlqpKqQn8VQRtekiSx8/edit#gid=1075670354).

### PR DESCRIPTION
**Pull request by Jack-0-0**
Closes #192.

For the `DiscourseAnalysis` class and related files, this PR:
  * Splits `load_documents` function into `load_documents` and `load_metadata`
  * Updates functions to work without metadata
  * Removes `view_mentions` and related functions as they perform the same function as `view_collocations`
  * Fixes function `noun_chunks_all_years` by changing instances of `_self_noun_chunks_all_years` to `_noun_chunks_all_years`
  * Fixes function `get_flat_sentences` by removing not needed for loop
  * Adds functionality to make general phrase patterns dictionary using the search terms (function `make_phrase_patterns`)
  * Adds functionality to create a BERTopic model
  * Fixes `subset_documents` function and makes `document_text` first subset on the search terms before required and banned terms
  * Updates `plot_document_mentions` function to `plot_mentions` which can now plot both document and sentence mentions
  
The notebook demoing the functionality can be found [here](https://colab.research.google.com/drive/1Mjtws6Aec9dxT8rpMHnK2P2yFPXY0DQ7?usp=sharing).

---

Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [ ] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained the feature in this PR or (better) in `output/reports/`
- [x] I have requested a code review

### CODE REVIEW

File: innovation_sweet_spots/utils/pd/pd_analysis_utils.py
Comment (beingkk) [inline innovation_sweet_spots/utils/pd/pd_analysis_utils.py]: Shouldn't the phrases be deduplicated first?

**Comment by beingkk** [comment]
Thank you very much @Jack-0-0 for this, it's really great! I've tested the code by running the Colab notebook (which worked very well for me).

### `plot_mentions` function
An issue I noticed with the `plot_mentions(use_documents=False)` function, when plotting number of sentences, was that it only uses the first column. Perhaps a quick fix can be to use a stacked bar chart and show all columns?

Screenshot 2022-11-10 at 17 48 04

### Adding an extra pattern
This is not a bug, but I wondered if we could add another pattern to the list of automatically generated patterns - I would like to see a variation of noun phrase where the search terms is in the front.

So, for example, at the moment, if I search for 'obesity' I'll have things like "adult obesity" and "childhood obesity".

However, I would also like to find things like "obesity crisis" or "obesity pandemic"

I think this could be relatively easily done by swapping a few lines for the very first pattern? eg, perhaps something like...
```
[ {'TEXT': 'obesity'},
  {'POS': 'NOUN'},
  {'POS': 'NOUN', 'OP': '?'},
]
```

You can also imagine examples with "food" like "food preparation", "food insecurity", "food poverty" etc.

### BERTopic visualisation: hover data
I realise this comes "out of the box" and perhaps we can't change anything, but maybe it's possible to:
- Add year to the hover-over data
- Break long sentences into multiple lines
However, it might be the case that it's not possible without changing BERTopic library itself (which we don't want to do here).

### BERTopic visualisation: topics
I'm a bit confused about why there are different number of topics in the first and second graph... does it just limit the bar chart graph to 8? It's also a bit annoying that the colours don't match between both, but that's a very minor issue.

Screenshot 2022-11-10 at 18 02 53

Screenshot 2022-11-10 at 18 03 08

**Review by beingkk — CHANGES_REQUESTED** [review]
Thanks so much again, I'm really happy about this PR - please see just a few small things I've commented above.

Once you make the changes, please don't merge it yet into dev, as I'm still working on finalising the public discourse analysis using the old utils. Once I'm done with my analysis, I will merge it and check that my code is not broken.

**Comment by Jack-0-0** [comment]
> ### `plot_mentions` function
> An issue I noticed with the `plot_mentions(use_documents=False)` function, when plotting number of sentences, was that it only uses the first column. Perhaps a quick fix can be to use a stacked bar chart and show all columns?
> 
I have updated `plot_mentions` to use a stacked bar chart now. If you run `plot_mentions(use_documents=False)`, it will produce a stacked bar chart.

Screenshot 2022-11-12 at 21 40 16


If you run `plot_mentions(use_documents=True)`, it will produce a regular bar chart as `document_mentions` only shows total mentions, it does not separate the mentions by different search terms.

**Comment by Jack-0-0** [comment]
> ### BERTopic visualisation: topics
> I'm a bit confused about why there are different number of topics in the first and second graph... does it just limit the bar chart graph to 8? It's also a bit annoying that the colours don't match between both, but that's a very minor issue.
> 

`visualize_barchart` has a parameter `top_n_topics` which defaults to 8.

I have set the notebook cell to now use `visualize_barchart(top_n_topics=len(set(topic_model.topics_)))` so that it shows all the topics.

I couldn't see any options in BERTopic to change the colour palette.

**Comment by Jack-0-0** [comment]
> ### BERTopic visualisation: hover data
> I realise this comes "out of the box" and perhaps we can't change anything, but maybe it's possible to:
> 
> * Add year to the hover-over data
> * Break long sentences into multiple lines
>   However, it might be the case that it's not possible without changing BERTopic library itself (which we don't want to do here).
> 

I couldn't see these options in BERTopic.

However, you can adjust the size of the plot with the parameters `width` and `height`. When the plot is larger it allows for longer sentences to be displayed. I have increased the size of the plot in the demo notebook and added a note relating to these parameters for users.

**Comment by Jack-0-0** [comment]
> ### Adding an extra pattern
> This is not a bug, but I wondered if we could add another pattern to the list of automatically generated patterns - I would like to see a variation of noun phrase where the search terms is in the front.
> 
> So, for example, at the moment, if I search for 'obesity' I'll have things like "adult obesity" and "childhood obesity".
> 
> However, I would also like to find things like "obesity crisis" or "obesity pandemic"
> 
> I think this could be relatively easily done by swapping a few lines for the very first pattern? eg, perhaps something like...
> 
> ```
> [ {'TEXT': 'obesity'},
>   {'POS': 'NOUN'},
>   {'POS': 'NOUN', 'OP': '?'},
> ]
> ```
> 
> You can also imagine examples with "food" like "food preparation", "food insecurity", "food poverty" etc.
> 

I have added the pattern as suggested.
It is named `f"{search_term}_term_noun"` in the phrase patterns dictionary.

**Comment by Jack-0-0** [comment]
Thanks for the review @beingkk. I have made changes relating to your comments.
Let me know when you're happy for me to merge 🙂

**Review by beingkk — APPROVED** [review]
Thanks so much Jack, looks good to merge!

**Pull request merged**
